alf/networks/s5/utils.py

```diff
@@ -83,248 +81,263 @@ def make_DPLR_HiPPO(N):
+            -grad_lambda_imag,
+            mask=mask)
+
+    class _ssm_forward(torch.autograd.Function):
+        # TODO use @triton.autotune to choose the best BLOCK_SIZE
+        # BLOCK_SIZE = 128 seems work well for 3090
+        BLOCK_SIZE = 128
+
+        @staticmethod
+        def forward(ctx, s, x, Lambda):
+            assert s.is_contiguous() and x.is_contiguous(
+            ) and Lambda.is_contiguous()
+            length, batch_size, dim = x.shape
+            n = batch_size * dim
+            y = torch.zeros_like(x)
+            grid = lambda meta: (triton.cdiv(n, meta['BLOCK_SIZE']), )
+
+            if Lambda.dtype == torch.complex64:
+                diag_ssm_forward_kernel_complex[grid](
+                    torch.view_as_real(s), torch.view_as_real(x),
+                    torch.view_as_real(y), torch.view_as_real(Lambda), length,
+                    batch_size, dim, _ssm_forward.BLOCK_SIZE)
+            elif Lambda.dtype.is_floating_point:
+                diag_ssm_forward_kernel[grid](s, x, Lambda, y, length,
+                                              batch_size, dim,
+                                              _ssm_forward.BLOCK_SIZE)
+            else:
+                raise ValueError("Unsupported dtype: %s" % Lambda.dtype)
+            ctx.save_for_backward(s, y, Lambda)
+            return y
+
+        @staticmethod
+        def backward(ctx, grad_y):
+            s, y, Lambda = ctx.saved_tensors
+            length, batch_size, dim = y.shape
+            grad_y = grad_y.contiguous()
+            n = batch_size * dim
+            grad_s = torch.empty_like(s)
+            grad_x = torch.empty_like(grad_y)
+            # Here grad_lambda stores the gradients of Lambda for each sample
+            # in the batch. We will sum them up after the kernel finishes.
+            grad_lambda = torch.empty_like(s)
+            grid = lambda meta: (triton.cdiv(n, meta['BLOCK_SIZE']), )
+            if Lambda.dtype == torch.complex64:
+                diag_ssm_backward_kernel_complex[grid](
+                    torch.view_as_real(s), torch.view_as_real(Lambda),
+                    torch.view_as_real(y), torch.view_as_real(grad_s),
+                    torch.view_as_real(grad_x),
+                    torch.view_as_real(grad_lambda),
+                    torch.view_as_real(grad_y), length, batch_size, dim,
+                    _ssm_forward.BLOCK_SIZE)
+            else:
+                diag_ssm_backward_kernel[grid](
+                    s, Lambda, y, grad_s, grad_x, grad_lambda, grad_y, length,
+                    batch_size, dim, _ssm_forward.BLOCK_SIZE)
+            return grad_s, grad_x, grad_lambda.sum(dim=0)
+
+    diag_ssm_forward_triton = _ssm_forward.apply
+
+except ImportError:
+    from alf.utils.common import warning_once
+
+    def diag_ssm_forward_triton(s, x, Lambda):
+        warning_once("Triton is not installed. Using slow diag_ssm_forward.")
+        return diag_ssm_forward_slow(s, x, Lambda)
```

Comment:
This has the disadvantage of falsely pass unit test when slow and triton paths actually differ.
Maybe try something like this:
https://github.com/HorizonRobotics/alf/commit/2ac807c39a1ca88bf36a51386b9630619966d1fe

Reply:
If triton is installed, it is always tested. If triton is not installed, then there is no way to test.

Reply:
If say our CI docker image accidentally removes triton, then the unit tests will silently pass, instead of raise an error.

Reply:
Our CI docker only runs in CPU.